### ISSUE TITLE
CURIEs are not added to the root resource

### ISSUE THREAD

**Issue by v3rm0n** [spring-projects/spring-hateoas#316]
In case of collections, "curies" link is added to all embedded resources. The spec says: "CURIEs are established within a HAL document via a set of Link Objects with the relation type "curies" on the root Resource Object"

**Comment by sesto** [comment]
+1. Moreover, if embedded resources have only self-links, no curies are added at all. Curie name space "ab" is added, though:
{
  "_links": {
    "self": {
      "href": "blah"
    }
  },
  "_embedded": {
    "ab:customerResourceList": [
      {
        "customerId": 12345,
        "_links": {
          "self": {
            "href": "Blah"
          }
        }
      },
      {
        "customerId": 9876,
        "_links": {
          "self": {
            "href": "Blah1"
          }
        }
      }
    ]
  }
}

**Comment by otnipesoj** [comment]
sesto, I have the same problem (described at http://stackoverflow.com/questions/28830396/curies-in-spring-hateoas/28832622#28832622) and I'm still waiting for a solution.

**Comment by olivergierke** [comment]
The reason no curie link shows up here is that curies are defined on links, not relations that appear in `_embedded`. This means, as long as there is no *link* that needs to be curied, not `curies` link is rendered in turn.

I can see that this might be desirable, so I'll keep this around as a suggested improvement.

**Comment by slinstaedt** [comment]
I also ran into issues with missing curies rendering, even though using curried links:
```
@RestController
public class HelloController {

    @RequestMapping(path = "/hello", produces = "application/hal+json")
    public ResponseEntity<EntityModel<Greeting>> hello(@RequestParam String name) {
        Greeting greeting = new Greeting("Hello " + name);
        Link world = linkTo(methodOn(HelloController.class).hello("World")).withRel("a:world");
        return ResponseEntity.ok(EntityModel.of(greeting).add(world));
    }
}

@Configuration
@EnableWebMvc
@EnableHypermediaSupport(type = {HypermediaType.HAL})
class Config {

    @Bean
    public CurieProvider curieProvider() {
        return new DefaultCurieProvider(Map.of( //
                "a", UriTemplate.of("https://aaa/rels/{rel}"), //
                "b", UriTemplate.of("https://bbb/rels/{rel}") //
        ), null);
    }
}

class Greeting {

    private String message;

    public Greeting(String message) {
        this.message = message;
    }

    public String getMessage() {
        return message;
    }
}
```

Using latest spring-boot-starter-parent:2.4.0

**Comment by odrotbohm** [comment]
@slinstaedt – Any chance you throw this into an executable project ideally with a test case invoking the controller?

**Comment by slinstaedt** [comment]
Sure. I just hoped you won't ask for that.
[curie-test.tar.gz](https://github.com/spring-projects/spring-hateoas/files/5577568/curie-test.tar.gz)

**Comment by odrotbohm** [comment]
I've filed #1399 as this seems to be a different issue than the one originally described. Fix coming in a bit.

**Comment by odrotbohm** [comment]
Closing this one as what's been asked for here originally has already been implemented with #320.
